src/pass/zero_elimination.cc

```diff
@@ -0,0 +1,1805 @@
+        } else {
+          shift = Select::make(best_lower <= -bnd.coef,
+                               best_lower / bnd.coef,
+                               (best_lower + bnd.coef - 1)/bnd.coef);
+        }
+        shift = SuperSimplify(shift, vranges);
+
+        Expr diff = SuperSimplify(best_diff_over / bnd.coef, vranges);
+
+        if (is_const_int(diff, 0)) {
+          // Don't create an itervar, just replace it everywhere with its min
+          res.old_to_new.Set(var, shift);
+        } else {
+          res.old_to_new.Set(var, new_var + shift);
+          // Note that we are substituting old with new, so best_lower contains new var,
+          // that is we have to substitute new with old in best_lower here
+          res.new_to_old.Set(new_var,
+                             SuperSimplify(var - Substitute(shift, res.new_to_old), vranges));
+
+          new_var_intsets[new_var.get()] = IntSet::interval(make_zero(new_var.type()), diff);
+
+          // Add the new var to the resulting axis
+          auto range = Range(make_zero(new_var.type()), SuperSimplify(diff + 1, vranges));
+          res.axis.push_back(new_var);
+          res.ranges.Set(new_var, range);
+          vranges.Set(new_var, range);
+        }
+      }
+    }
+  }
+
+  // Add the original conditions (with variables substituted) to the resulting conditions
+  for (const Expr& old_cond : solved_system.as_conditions()) {
+    res.conditions.push_back(SuperSimplify(Substitute(old_cond, res.old_to_new), vranges));
+  }
+
+  return res;
+}
+
+// Use the condition of a reduction op to simplify its domain (axis)
+Expr SimplifyReductionDomain(const Expr& expr, const Map<Var, Range>& outer_vranges) {
+  if (const Reduce* red = expr.as<Reduce>()) {
+    Map<Var, Range> vranges = Merge(outer_vranges, IterVarsToMap(red->axis));
+    auto res = SimplifyDomain(red->condition, IterVarsToVars(red->axis),
+                              Merge(outer_vranges, IterVarsToMap(red->axis)));
+
+    Array<Expr> new_source;
+    for (const Expr& src : red->source) {
+      new_source.push_back(Substitute(src, res.old_to_new));
+    }
+
+    Array<IterVar> new_axis = IterVarsFromMap(res.axis, res.ranges, kCommReduce);
+
+    // Perform simplification mainly to remove a possibly empty reduction.
+    return Simplify(Reduce::make(red->combiner, new_source, new_axis,
+                                 All(res.conditions), red->value_index));
+  } else {
+    return expr;
+  }
+}
+
+// Extract the given expr under the given condition as a separate tensor if the volume of the
+// extracted tensor will be less than the volume of the outer_axis
+Expr ExtractAsTensorMaybe(const Expr& e, const Expr& cond,
+                          const Array<Var>& outer_axis,
+                          const Map<Var, Range>& vranges) {
+  // TODO(sgrechanik-h): We don't use divmod elimination here because of some performance problems
+  auto res = SimplifyDomain(cond, outer_axis, vranges, false);
```

Comment:
SimplifyDomain iterates over outer_axis in reverse order at line 1297 (as described by equations 22-26 in the paper?).  As such, res.axis and res.conditions are in opposite orders.  (conditions are added at line 1394 in the original order)  This can result in a unnecessary transpose copy by line 1468 when trying to create a TensorFromExpr with res.axis which contain axes in reverse order from that of new_expr.   Should a simple fix in SimplifyDomain ensure original order of simplified axes in res (i.e. not reverse), a unnecessary copy can still result from TensorFromExpr.   Any thoughts?  Tensor copy can be costly to performance.

Reply:
@kaitingwang I think it makes sense to reverse the resulting `res.axis` to match the original order of variables, I'll fix it. However, although it helps in some cases, the real problem is in the ExtractAsTensorMaybe function which doesn't check if the expression is costly enough to make the extraction beneficial. Since "costly enough" is rather difficult to define precisely, I think, the proper solution is to rely on schedules (`compute_inline` should help, I guess). However, this is quite naive (especially considering that we don't have an autoscheduler yet), so I think I'll implement some heuristics to prevent such extractions in some obvious cases.

Reply:
Indeed, in cases where new_expr is a computational intensive expression,  computing it once and memoize the result for later use is wise.  This is in the case where the uses may require recomputing the same computational intensive expressions multiple times.    Thanks for the fix.